modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/rbac/CacheRBACImpl.java

```diff
@@ -17,37 +21,72 @@
     private IAuthStoreService authStoreService;
     private RBACService rbacService;
-    
-    public CacheRBACImpl(IAuthStoreService authStoreService, RBACService rbacService) {
+
+    public CacheRBACImpl(
+        IDynamicStatusStoreService dssService,
+        IAuthStoreService authStoreService,
+        RBACService rbacService
+    ) {
+        this.dssService = dssService;
         this.authStoreService = authStoreService;
         this.rbacService = rbacService;
     }
 
     @Override
     public synchronized void addUser(String loginId, List<String> actionIds) throws RBACException {
-        // Do nothing for now...
+        try {
+            // Create a property in the DSS of the form:
+            // dss.rbac.<loginId>.actions = <comma-separated action IDs>
+            String commaSeparatedActionIds = String.join(",", actionIds);
+            String actionsKey = loginId + ACTIONS_PROPERTY_SUFFIX;
+            dssService.put(actionsKey, commaSeparatedActionIds, CACHED_ACTIONS_TIME_TO_LIVE_SECS);
+        } catch (DynamicStatusStoreException e) {
+            throw new RBACException("Failed to cache user actions", e);
+        }
     }
 
     @Override
     public synchronized boolean isActionPermitted(String loginId, String actionId) throws RBACException {
-        IUser user = getUserFromAuthStore(loginId);
-        String userRoleId = user.getRoleId();
-        if (userRoleId == null) {
-            userRoleId = rbacService.getDefaultRoleId();
-            user.setRoleId(userRoleId);
-        }
-        Role userRole = rbacService.getRoleById(userRoleId);
+        boolean isActionPermitted = false;
+        try {
+            String userActionsKey = loginId + ACTIONS_PROPERTY_SUFFIX;
+            String commaSeparatedUserActions = dssService.get(userActionsKey);
+
+            List<String> userActions = new ArrayList<>();
+            if (commaSeparatedUserActions == null) {
+                // Cache miss, so get the user's actions from the auth store
+                userActions = getUserActionsFromAuthStore(loginId);
 
-        // Check if the user is allowed to perform the given action
-        boolean isActionPermitted = userRole.getActionIds().contains(actionId);
+                // Add this user to the cache
+                addUser(loginId, userActions);
+            } else {
+                userActions = Arrays.asList(commaSeparatedUserActions.split(","));
+            }
+
+            // Check if the user is allowed to perform the given action
+            isActionPermitted = userActions.contains(actionId);
+        } catch (DynamicStatusStoreException e) {
+            throw new RBACException("Error occurred when accessing the DSS", e);
+        }
         return isActionPermitted;
     }
 
     @Override
     public synchronized void invalidateUser(String loginId) throws RBACException {
-        // Do nothing for now...
+        try {
+            String userActionsKey = loginId + ACTIONS_PROPERTY_SUFFIX;
```

Comment:
duplicate code calculating the dss key ? Logic would be better only in one place.

Reply:
Done in https://github.com/galasa-dev/galasa/pull/162/commits/011d27b7bf48c4812b532fdbf9eb7eb2c9a7be3c